Adding all the different types of chips, along with their complex behaviors

gspencergoog opened the pull request:
This implements various different chip types: There are now the `InputChip`, `ChoiceChip`, `FilterChip`,  `ActionChip`, and the original  vanilla `Chip`, all of which have different uses.  They can be customized in a number of ways. See their docs for more information.

![image](https://user-images.githubusercontent.com/8867023/38159425-7b529008-345d-11e8-8575-dfa2860d8de2.png)

Also fixes #16083.

gspencergoog committed:
Adding all the different types of chips, along with their complex behaviors.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+/// Input chips represent a complex piece of information, such as an entity
+/// (person, place, or thing) or conversational text, in a compact form.
+///
+/// Input chips can be selected by setting [onSelected], deleted by setting
```
I don't think you mean that input chips can be selected by setting their onSelected callback.  And etc.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+/// Input chips work together with other UI elements. They can appear:
+///
+///  * In a [Wrap] widget.
+///  * In a horizontally scrollable list
```
Might as well provide some links here: ... like a [ListView] whose `scrollDirection` is [Axis.horizontal].

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+/// and they can have a leading icon (see [avatar]) and a trailing icon
+/// ([deleteIcon]). Colors and padding can be customized.
+///
+/// Requires one of its ancestors to be a [Material] widget. The [label],
```
We usually put the "must not be null" stuff in the constructor doc.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+///   ),
+///   label: new Text('Aaron Burr'),
+///   onPressed: () {
+///     print("I am the one thing in life.");
```
Single quotes

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+///  * [CircleAvatar], which shows images or initials of people.
+///  * <https://material.google.com/components/chips.html>
+class InputChip extends StatelessWidget {
+  const InputChip({
```
Constructor doc

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+class Chip extends StatelessWidget {
+  /// Creates a material design chip.
+  ///
+  /// The [label] and [border] arguments may not be null.
```
We've been consistently saying "must not" for boilerplate like this

HansMuller commented:
I haven't really made it to the heart of this PR yet. The feedback so far is trivial stuff, mostly just to mark my progress.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+  /// Called when the user taps the chip.
+  ///
+  /// If [onPressed] is set, then this callback will be called when the user
+  /// taps on the label or avatar parts of the chip. If [onPressed] is null,
```
See earlier comment about onPressed==null and disabled.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+    return _wrapWithTooltip(
+      new InkResponse(
+        onTap: widget.isEnabled ? widget.onDeleted : null,
+        child: new Container(
```
Don't need this Container?

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+              color: widget.deleteIconColor ?? theme.iconTheme.color,
+              opacity: _kDeleteIconOpacity,
+            ),
+            child: widget.deleteIcon ?? const Icon(Icons.cancel, size: _kDeleteIconSize),
```
Could just have a _kDefaultDeleteIcon

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
     assert(debugCheckHasMaterial(context));
     final ThemeData theme = Theme.of(context);
+    final TextDirection textDirection = Directionality.of(context);
+    _reactToChanges();
```
Don't need this, override didUpdateWidget() instead

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+      style: widget.labelStyle ??
           theme.textTheme.body2.copyWith(
-            color: theme.primaryColorDark.withAlpha(_kTextLabelAlpha),
+            color: Colors.black.withAlpha(_kTextLabelAlpha),
```
_kTextLabelColor

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+          onTap: canTap ? _handleTap : null,
+          onTapDown: canTap ? _handleTapDown : null,
+          onTapCancel: canTap ? _handleTapCancel : null,
+          child: new Container(
```
Instead of this container, just set the Material's color

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -344,9 +1579,11 @@ class _ChipRenderTheme {
-        typedOther.container == container &&
         typedOther.padding == padding &&
-        typedOther.labelPadding == labelPadding;
+        typedOther.labelPadding == labelPadding &&
```
The latest style for this sort of thing puts the `&&` on the left so that all of the clauses line up:
```
    return typedOther.foo == foo
        && typedOther.bar == bar
        && typedOther.baz == baz

gspencergoog replied:
Just to be clear, I'll make these changes before I submit, but I don't want to do pure formatting changes that don't match what dartfmt does until I'm done with all the other changes.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -366,14 +1605,60 @@ class _RenderChip extends RenderBox {
-    _tap = new TapGestureRecognizer(debugOwner: this)
-      ..onTapDown = _handleTapDown
-      ..onTap = _handleTap;
+    if (selectController != null) {
```
Shouldn't all of this be handled by an AnimatedBuilder in RawChip?

gspencergoog committed:
Adding more tests

gspencergoog committed:
Review changes

gspencergoog commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+  /// delete icon will be shown.
+  final String deleteButtonTooltipMessage;
+
+  /// The border to draw around the chip.
```
It's called border here because that's what I called it when I added it originally.  I'm happy to change it, but it's a breaking change.

gspencergoog committed:
Move animations to RawChip

gspencergoog commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+    selectionFade = new CurvedAnimation(
+      parent: selectController,
+      curve: Curves.fastOutSlowIn,
+    )..addListener(() {
```
Done!  Thanks, that really helped a lot.

gspencergoog commented on packages/flutter/lib/src/material/chip.dart:
````diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+/// ```dart
+/// bool _value = false;
+///
+/// new ChoiceChip(
````
Thanks, that works better.

gspencergoog commented on packages/flutter/lib/src/material/chip.dart:
````diff
@@ -10,154 +10,1371 @@ import 'package:flutter/painting.dart';
+/// ```
+///
+/// See also:
+///
````
Done.

gspencergoog commented:
OK, I think I've addressed most of your comments, and I think things are looking pretty good, modulo the rest of the tests. I want to add some more examples, but those can be in another PR.

gspencergoog commented on packages/flutter/lib/src/material/ink_well.dart:
```diff
@@ -218,6 +220,13 @@ class InkResponse extends StatefulWidget {
   /// Called when the user taps this part of the material.
   final GestureTapCallback onTap;
 
+  /// Called when the user taps down this part of the material.
```
Done. https://github.com/flutter/flutter/pull/16190

gspencergoog committed:
Move InkWell changes to another PR

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1373 @@ import 'package:flutter/painting.dart';
+  /// taps on the label or avatar parts of the chip. If [onPressed] is null,
+  /// then the chip will be disabled.
+  ///
+  /// The [onPressed] and [SelectableChipAttributes.onSelected] callbacks must
```
This point probably belongs in the implementing class

gspencergoog replied:
OK, I moved it to the constructors for InputChip and RawChip (the only ones where it applies).

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1373 @@ import 'package:flutter/painting.dart';
+/// }
+///
+/// class _MyThreeOptionsState extends State<MyThreeOptions> {
+///   int _value;
```
Needs an initial value

gspencergoog replied:
Done.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1373 @@ import 'package:flutter/painting.dart';
+/// Requires one of its ancestors to be a [Material] widget. The [label]
+/// and [border] arguments must not be null.
+///
+/// ## Sample code
```
This doesn't look like a FilterChip example

gspencergoog replied:
That's because it isn't. Fixed.

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1373 @@ import 'package:flutter/painting.dart';
+                          theme.textTheme.body2.copyWith(
+                            color: Colors.black.withAlpha(_kTextLabelAlpha),
+                          ),
+                      child: widget.label),
```
close parent on the next line

gspencergoog replied:
Done.

HansMuller commented on packages/flutter/test/material/chip_test.dart:
```diff
@@ -212,14 +230,16 @@ void main() {
         new OverlayEntry(
           builder: (BuildContext context) {
             return new Material(
               child: new Center(
```
This Center widget probably isn't needed for a correct test and it may be introducing some extra floating point variability in the chips' locations.

gspencergoog replied:
Good point.  I'll take it out.

HansMuller reviewed (commented):
This looks great. The only issue of consequence is how the FooAttribute should define or just explain their default values.

gspencergoog commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1373 @@ import 'package:flutter/painting.dart';
+  ///     const Actor('James Madison', 'JM'),
+  ///   ];
+  ///
+  ///   Iterable<Widget> get actorWidgets sync* {
```
Yeah, I liked that too.

gspencergoog committed:
Review Changes

HansMuller commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1373 @@ import 'package:flutter/painting.dart';
+
+  /// The border to draw around the chip.
+  ///
+  /// Defaults to a [StadiumBorder]. Must not be null.
```
[Discusssed with @gspencergoog  and @zanderso] There's no simple tidy way for the properties in the abstract classes to define a default value. So each abstract class will have to include a note about what "default value" means vis the read-only properties.

Hixie replied:
See also this approach which you can use for sharing docs while having per-instance details:
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/navigator.dart#L606
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/navigator.dart#L1346

HansMuller commented:
LGTM

Hixie commented:
This added three undocumented members, FWIW.

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1447 @@ import 'package:flutter/painting.dart';
+///  * [FilterChip], uses tags or descriptive words as a way to filter content.
+///  * [ActionChip], represents an action related to primary content.
+///  * <https://material.google.com/components/chips.html>
+abstract class ChipAttributes {
```
for classes you don't want people to extend (i.e. mixins and interfaces), include the following boilerplate code:
```dart
  // This class is intended to be used as an interface, and should not be
  // extended directly.
  factory ChipAttributes._() => null;
```

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -10,154 +10,1447 @@ import 'package:flutter/painting.dart';
+  /// The border to draw around the chip.
+  ///
+  /// Defaults to a [StadiumBorder]. Must not be null.
+  ShapeBorder get border;
```
normally we call this `shape`, not `border`; was this an intentional choice?

gspencergoog replied:
Well... yes?  It was what I originally named it when I created the StadiumBorder, and so it's in the Chip API, so I didn't want to change it for this.  If you're OK with breaking the few people who may have discovered this feature, I'm happy doing that too.

Hixie replied:
Ah, makes sense.

Sure, let's change it. We probably have like 3 people using it right now. Long term benefit of being consistent everywhere probably outweighs that.

Hixie commented:
I'll add docs for the undocumented members.

I love RawChip. It's like the swiss army knife of chips.
